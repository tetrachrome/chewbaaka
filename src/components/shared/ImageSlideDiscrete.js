/**
 * ImageSlideDiscrete.js
 * Chewbaaka
 *
 * Author   : Tomiko
 * Created  : Jul 04, 2020
 * Updated  : Jul 05, 2020
 */

/**
 * ImageSlideDiscrete
 *
 * ImageSlideDiscrete is a component that allows user to
 * specify a set of images to display along the progress
 * indicated on a progress bar. A good example application
 * of this component could be to use a set of images to
 * illustrate changes along a timeline.
 *
 * Props:
 *  - `images`: An array of images to display within the component.
 *
 *  - `labels` (optional): An array of label captions to display inside the progress indicator.
 *    If specified, this array must have the same length as the `images` prop specified.
 *
 *  - `percentages` (optional): An array of increment percentages for the progress indicator.
 *    If specified, this array must have the same length as the `images` prop specified.
 *
 *  - `title` (optional): The title of the component.
 *
 *  - `decrementButtonLabel`: The label used for the button for decrementing the progress.
 *
 *  - `incrementButtonLabel`: The label used for the button for incrementing the progress.
 *
 *  - `controlButtonsAlignment` (optional): Specify the alignment of the button controls
 *    w.r.t. the entire control.
 */

import React from 'react';

import "semantic-ui-css/semantic.min.css";

import { Button, Container, Header, Icon, Image, Progress } from 'semantic-ui-react'

class ImageSlideDiscrete extends React.Component {

  // TODO: Style this component.

  constructor(props) {
    super(props);
    this.state = {
      percent: 0,
      index: 0
    }
    this.handleDecrementButtonClick = this.handleDecrementButtonClick.bind(this);
    this.handleIncrementButtonClick = this.handleIncrementButtonClick.bind(this);
  }

  handleDecrementButtonClick() {
    const increment = 100 / Math.max(this.props.images.length, 1);

    this.setState(function setState(prevState) {
      const index = Math.max(prevState.index - 1, 0);
      const percent = this.props.percentages ? this.props.percentages[index] : Math.max(prevState.percent - increment, 0);
      return {
        percent: percent,
        index: index,
      }
    });
  }

  handleIncrementButtonClick() {
    const increment = 100 / Math.max(this.props.images.length, 1);

    this.setState(function setState(prevState) {
      const index = Math.min(prevState.index + 1, this.props.images.length-1);
      const percent = this.props.percentages ? this.props.percentages[index] : Math.min(prevState.percent + increment, 100);
      return {
        percent: percent,
        index: index,
      }
    });
  }

  render() {
    const controlButtonsAlignment = this.props.controlButtonsAlignment? this.props.controlButtonsAlignment : "center";

    const labels = this.props.labels ? this.props.labels : [];

    return (
      <div data-testid="ImageSlideComponentTestId">
        <Header as='h1' data-testid="ImageSlideComponentTitlePartTestId">{this.props.title}</Header>
        <Container>
          <Image src={this.props.images[this.state.index]} fluid data-testid="ImageSlideComponentImgPartTestId" />
        </Container>
        <Progress size="tiny" percent={this.state.percent} indicating data-testid="ImageSlideComponentProgressIndicatorPartTestId" >
          {labels[this.state.index]}
        </Progress>
        <Container textAlign={controlButtonsAlignment}>
          <Button icon labelPosition='left' onClick={this.handleDecrementButtonClick} data-testid="ImageSlideComponentDecrementButtonPartTestId" >
            {this.props.decrementButtonLabel}
            <Icon name='left arrow' />
          </Button>
          <Button icon labelPosition='right' onClick={this.handleIncrementButtonClick} data-testid="ImageSlideComponentIncrementButtonPartTestId">
            {this.props.incrementButtonLabel}
            <Icon name='right arrow' />
          </Button>
        </Container>
      </div>
    )
  }
}

export default ImageSlideDiscrete;
