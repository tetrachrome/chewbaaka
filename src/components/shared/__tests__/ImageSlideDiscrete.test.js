/**
 * ImageSlideDiscrete.test.js
 * Chewbaaka
 *
 * Author   : Tomiko
 * Created  : Jul 04, 2020
 * Updated  : Jul 08, 2020
 */

import React from 'react';

import { render, screen } from '@testing-library/react';

import renderer from 'react-test-renderer';

import ImageSlideDiscrete from '../ImageSlideDiscrete';

import image01 from '../assets/cheetah-conservation-fund-logo.jpg'
import image02 from '../assets/cheetah-conservation-fund-30-logo.svg'

const title = "Cheetah evolution";

test('renders ImageSlideDiscrete component', () => {
  render(
          <ImageSlideDiscrete images={[image01, image02]}
          labels={["image01", "image02"]}
          percentages={[0, 20,100]} title={title}
          decrementButtonLabel="backward"
          incrementButtonLabel="forward"/>
  );

  const imageSlideComponent = screen.getByTestId("ImageSlideDiscreteComponentTestId");
  expect(imageSlideComponent).toBeInTheDocument();

  // Tests title component part.
  const imageSlideComponentTitlePart = screen.getByTestId("ImageSlideDiscreteComponentTitlePartTestId");
  expect(imageSlideComponentTitlePart).toBeInTheDocument();
  expect(imageSlideComponentTitlePart.textContent).toBe(title);

  // Tests img component part.
  const imageSlideComponentImgPart = screen.getByTestId("ImageSlideDiscreteComponentImgPartTestId");
  expect(imageSlideComponentImgPart).toBeInTheDocument();

  // Tests progress indicator component part.
  const imageSlideComponentProgressIndicatorPart = screen.getByTestId("ImageSlideDiscreteComponentProgressIndicatorPartTestId");
  expect(imageSlideComponentProgressIndicatorPart).toBeInTheDocument();

  // Tests decrement button component part.
  const imageSlideComponentDecrementButton = screen.getByTestId("ImageSlideDiscreteComponentDecrementButtonPartTestId");
  expect(imageSlideComponentDecrementButton).toBeInTheDocument();

  // Tests increment button component part.
  const imageSlideComponentIncrementButton = screen.getByTestId("ImageSlideDiscreteComponentIncrementButtonPartTestId");
  expect(imageSlideComponentIncrementButton).toBeInTheDocument();
});

test('ImageSlideDiscrete component snapshot', () => {
  const tree = renderer
    .create(
      <ImageSlideDiscrete
        images={[image01, image02]}
        labels={["image01", "image02"]}
        percentages={[0, 20,100]} title={title}
        decrementButtonLabel="backward"
        incrementButtonLabel="forward"
      />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
