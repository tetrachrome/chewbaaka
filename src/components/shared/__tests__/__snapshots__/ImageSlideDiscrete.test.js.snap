// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageSlideDiscrete component snapshot 1`] = `
<div
  data-testid="ImageSlideDiscreteComponentTestId"
>
  <h1
    className="ui header"
    data-testid="ImageSlideDiscreteComponentTitlePartTestId"
  >
    Cheetah evolution
  </h1>
  <div
    className="ui container"
  >
    <img
      className="ui fluid image"
      data-testid="ImageSlideDiscreteComponentImgPartTestId"
      src="cheetah-conservation-fund-logo.jpg"
    />
  </div>
  <div
    className="ui tiny progress"
    data-percent={0}
    data-testid="ImageSlideDiscreteComponentProgressIndicatorPartTestId"
  >
    <div
      className="bar"
      style={
        Object {
          "width": "0%",
        }
      }
    />
    <div
      className="label"
    >
      image01
    </div>
  </div>
  <div
    className="ui center aligned container"
  >
    <button
      className="ui icon left labeled button"
      data-testid="ImageSlideDiscreteComponentDecrementButtonPartTestId"
      onClick={[Function]}
    >
      backward
      <i
        aria-hidden="true"
        className="left arrow icon"
        onClick={[Function]}
      />
    </button>
    <button
      className="ui icon right labeled button"
      data-testid="ImageSlideDiscreteComponentIncrementButtonPartTestId"
      onClick={[Function]}
    >
      forward
      <i
        aria-hidden="true"
        className="right arrow icon"
        onClick={[Function]}
      />
    </button>
  </div>
</div>
`;
